smart_control/dataset/dataset.py

```diff
@@ -1,96 +1,355 @@
 
-class SmartBuildingsDataset:
-  """Smart Buildings Dataset"""
-
-  def __init__(self, download=True):
-    self.partitions = {
-        "sb1": [
-            "2022_a",
-            "2022_b",
-            "2023_a",
-            "2023_b",
-            "2024_a",
-        ],
-    }
-    if download:
+# import pandas as pd
+
+
+DATA_DIR = os.path.join(ROOT_DIR, "data")
+
+VALID_BUILDING_PARTITIONS = {
+    "sb1": ["2022_a", "2022_b", "2023_a", "2023_b", "2024_a"]
+}
+
+
+class BuildingDataset:
+  """A helper class for handling the dataset for a specific building.
+
+  Args:
+    building_id (str): The identifier of the building (e.g. "sb1").
+    download (bool): Whether or not to download the dataset.
+
+  Examples:
+    ```python
+    ds = BuildingDataset(building_id="sb1", download=True)
+    ```
+  """
+
+  def __init__(self, building_id="sb1", download=True):
+    self.building_id = building_id
+
+    if self.building_id not in VALID_BUILDING_PARTITIONS:
+      raise ValueError("Invalid building: '{self.building_id}'.")
+
+    self.partition_ids = VALID_BUILDING_PARTITIONS[self.building_id]
+
+    if bool(download):
       self.download()
 
-  def download(self):
-    """Downloads the Smart Buildings Dataset from Google Cloud Storage."""
-    print("Downloading data...")
+  @property
+  def zip_filename(self):
+    return f"{self.building_id}.zip"
 
-    def download_file(url, timeout=200):
-      local_filename = url.split("/")[-1]
+  @property
+  def zip_url(self):
+    return (
+        "https://storage.googleapis.com/gresearch/smart_buildings_dataset/"
+        f"tabular_data/{self.zip_filename}"
+    )
 
+  @property
+  def zip_filepath(self):
+    return os.path.join(DATA_DIR, self.zip_filename)
+
+  @property
+  def building_dirpath(self):
+    return os.path.join(DATA_DIR, self.building_id)
+
+  @staticmethod
+  def _download_file_if_not_exists(url, timeout=60):
+    local_filename = url.split("/")[-1]
+    local_filepath = os.path.join(DATA_DIR, local_filename)
+
+    if os.path.isfile(local_filepath):
+      print("Using previously-downloaded data...")
+      print(os.path.abspath(local_filepath))
+    else:
+      print("Downloading data...")
+      print(url)
       with requests.get(url, stream=True, timeout=timeout) as r:
         r.raise_for_status()
-
-        with open(local_filename, "wb") as f:
+        with open(local_filepath, "wb") as f:
           for chunk in r.iter_content(chunk_size=8192):
             f.write(chunk)
 
-      return local_filename
+    return local_filepath
 
-    url = "https://storage.googleapis.com/gresearch/smart_buildings_dataset/tabular_data/sb1.zip"  # pylint: disable=line-too-long
-    download_file(url)
-    shutil.unpack_archive("sb1.zip", "sb1/")
+  def download(self):
+    """Downloads the building's dataset from Google Cloud Storage."""
+    self._download_file_if_not_exists(self.zip_url)
+    shutil.unpack_archive(self.zip_filepath, self.building_dirpath)
```

Comment:
Move this inside the download if not exists function, to prevent unnecessary unpacking/ unzipping.